common/utils.h

```diff
@@ -19,15 +19,13 @@ char *tal_hex(const tal_t *ctx, const tal_t *data);
 /* Allocate and fill a buffer with the data of this hex string. */
 u8 *tal_hexdata(const tal_t *ctx, const void *str, size_t len);
 
-/* Helper macro to extend tal_arr and return pointer new last element. */
-#if HAVE_STATEMENT_EXPR
-/* More efficient version calls tal_count() once */
-#define tal_arr_expand(p)					\
-	({ size_t n = tal_count(*p); tal_resize((p), n+1); *(p) + n; })
-#else
-#define tal_arr_expand(p) \
-	(tal_resize((p), tal_count(*(p))+1), (*p) + tal_count(*(p))-1)
-#endif
+/* Note: p is never a complex expression, otherwise this multi-evaluates! */
+#define tal_arr_expand(p, s)						\
+	do {								\
+		size_t n = tal_count(*(p));				\
+		tal_resize((p), n+1);					\
+		(*(p))[n] = (s);					\
+	} while(0)
```

Comment:
It's a standard C trick for when you want a multi-statement macro.  This turns it into a single statement, otherwise you can get strange results with `if (something) MACRO;` not doing what you expect.

Reply:
Strange, wouldn't surrounding with `{}` have the same effect, or are statement blocks not allowed in C?

Reply:
Not quite, since the user will add a ';' turning it into two statements.  Now, this normally doesn't matter, except for the dangling else rule which says it binds to the last 'if'.  So, this suddenly breaks:

`if (something) MACRO; else ...`